### PR TITLE
[CIVP-10806] Add ModelFuture

### PR DESCRIPTION
**Pull request by stephen-hoover**
This is the next step in adding the new modeling interface. This PR adds the new `civis.ml` namespace, and populates it with `ModelFuture`, a subclass of `CivisFuture`. The `ModelFuture` object knows how to find and download results from a CivisML model. It also transports errors from Civis and re-raises them in the local environment to make debugging easier.

### CODE REVIEW

**Comment by stephen-hoover** [comment]
@keithing , I've split this PR into two commits. The first copy-pastes already-reviewed code from the modeling project repository, and the second modifies it to live in the API client and to work with the functions I added in #68 . If you review the commits separately, you can look at the first one at a higher level and assume that the details work.

**Comment by stephen-hoover** [comment]
Haha. Our test environment is Python 3.4.2. I'll remove the type hints (those require Python >= 3.5.0).

File: civis/ml/_model.py [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+            # then there's no improvements to make on
+            # any existing exception.
+            pass
+        except KeyError:
+            # KeyErrors always represent a bug in the modeling code,
+            # but showing the resulting KeyError can be confusing and
+            # mask the real error.
+            warnings.warn("Received malformed metadata from Civis Platform. "
+                          "Something went wrong with execution. "
+                          "Please report this error.")
```
Comment (keithing) [inline civis/ml/_model.py]: What's going on here?  Can we directly tell if this issue has occurred? I would prefer not to ask users to be reporting errors if we can avoid it.

File: civis/ml/_model.py [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+        buf.seek(0)
+        file_id = cio.file_to_civis(buf, name=civis_fname, client=client)
+    else:
+        with open(X) as _fin:
+            file_id = cio.file_to_civis(_fin, name=civis_fname, client=client)
+
+    return file_id
+
+
+def _decode_train_run(train_job_id, train_run_id, client):
```
Comment (keithing) [inline civis/ml/_model.py]: I don't see where these private functions are being called: `_decode_train_run`, `_stash_local_data` and `_check_is_fit`.  Are they only used in testing?

File: civis/ml/_model.py [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+    """Download a Civis file using a reference on a previous run
+
+    Parameters
+    ----------
+    fname: str
+        Name of the Civis file you wish to retrieve
+    job_id, run_id: int
+        Job ID and run ID of the run holding a reference to this file
+    local_dir: str
+        Download the file `fname` from Civis Platform to this directory
```
Comment (keithing) [inline civis/ml/_model.py]: `client` param is missing.  Might not even need param docs on the private function here.

File: civis/ml/_model.py (hunk 1 above)
Comment (stephen-hoover) [inline civis/ml/_model.py]: That should never actually occur. If we don't handle it, then the user would get some garbled exception that won't make much sense. Also, it wouldn't necessarily need to halt execution of whatever this is running in. I wanted to let the user know that something was wrong without forcing a stop. Should I remove the "Please report this error." line and leave the rest? Do something else?

File: civis/ml/_model.py (hunk 2 above)
Comment (stephen-hoover) [inline civis/ml/_model.py]: Apologies. Those will be used in the next PR. I could remove them from this PR and let you review them separately if you prefer.

File: civis/ml/_model.py (hunk 3 above)
Comment (stephen-hoover) [inline civis/ml/_model.py]: I removed the parameters section of this docstring, since it's a private function.

**Comment by stephen-hoover** [comment]
@keithing , your questions made me realize that I'd forgotten the "Parameters" section for `ModelFuture`. I added that.

File: civis/ml/_model.py (hunk 2 above)
Comment (keithing) [inline civis/ml/_model.py]: Oh, no that's fine.  I thought that might be the case, but wanted to check.

File: civis/ml/_model.py (hunk 1 above)
Comment (keithing) [inline civis/ml/_model.py]: Okay, perfect. Let's remove that line, but we don't need to do anything else.

Comment (stephen-hoover) [inline civis/ml/_model.py]: Removed!

**Pull request merged**